server/src/main/java/io/crate/expression/scalar/cast/ImplicitCastFunction.java

```diff
@@ -53,22 +58,55 @@ public static void register(ScalarFunctionModule module) {
 
     private final Signature signature;
     private final Signature boundSignature;
+    @Nullable
+    private final DataType<?> targetType;
 
     private ImplicitCastFunction(Signature signature, Signature boundSignature) {
+        this(signature, boundSignature, null);
+    }
+
+    private ImplicitCastFunction(Signature signature, Signature boundSignature, @Nullable DataType<?> targetType) {
         this.signature = signature;
         this.boundSignature = boundSignature;
+        this.targetType = targetType;
+    }
+
+    @Override
+    public Scalar<Object, Object> compile(List<Symbol> args) {
+        assert args.size() == 2 : "number of arguments must be 2";
+        Symbol input = args.get(1);
+        if (input instanceof Input) {
+            String targetTypeValue = (String) ((Input<?>) input).value();
+            var targetTypeSignature = parseTypeSignature(targetTypeValue);
+            var targetType = targetTypeSignature.createType();
+            return new ImplicitCastFunction(
+                signature,
+                boundSignature,
+                targetType
+            );
+        }
+        return this;
     }
 
     @Override
     public Object evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<Object>[] args) {
-        var targetTypeSignature = parseTypeSignature((String) args[1].value());
-        var targetType = targetTypeSignature.createType();
+        assert args.length == 1 || args.length == 2 : "number of args must be 1 or 2";
+        if (targetType == null) {
+            var targetTypeSignature = parseTypeSignature((String) args[1].value());
+            var targetType = targetTypeSignature.createType();
+            return castToTargetType(targetType, args[0]);
+        } else {
+            return castToTargetType(targetType, args[0]);
+        }
+    }
+
+    private static Object castToTargetType(DataType<?> targetType, Input<Object> arg) {
         try {
-            return targetType.implicitCast(args[0].value());
+            return targetType.implicitCast(arg.value());
         } catch (ConversionException e) {
             throw e;
         } catch (ClassCastException | IllegalArgumentException e) {
-            throw new ConversionException(args[0].value(), targetType);
+            throw new ConversionException(arg.value(), targetType);
```

Comment:
If we pass in the value instead of the input, or store the value locally, we'd avoid a 2nd expensive `value()` call in case of an error.